### DISCUSSION TITLE
What's the different between PreUnifiedGCLogParser and UnifiedGCLogParser。

### DISCUSSION THREAD

**Discussion by ZhangShushu123** [microsoft/gctoolkit#178] [Q&A] [answered]
I find my gc log has the feature both PreUnifiedGCLog and UnifiedGCLog Parser, like below. How should I do with them?
> 2021-12-01T00:04:10.865+0800: 354090.931: [GC pause (G1 Evacuation Pause) (young), 0.0777177 secs]
2021-12-01T18:49:54.435+0800: 421634.502: Total time for which application threads were stopped: 0.0013231 seconds, Stopping threads took: 0.0001834 seconds
{Heap before GC invocations=5542 (full 2):
garbage-first heap   total 16777216K, used 5765808K [0x00000003c0000000, 0x00000003c0408000, 0x00000007c0000000)
  region size 4096K, 815 young (3338240K), 30 survivors (122880K)
 Metaspace       used 152940K, capacity 162166K, committed 162316K, reserved 1196032K
  class space    used 15321K, capacity 16962K, committed 17024K, reserved 1048576K
   [Parallel Time: 48.3 ms, GC Workers: 18]
      [GC Worker Start (ms): Min: 421634440.6, Avg: 421634440.7, Max: 421634440.9, Diff: 0.4]
      [Ext Root Scanning (ms): Min: 1.1, Avg: 1.5, Max: 3.5, Diff: 2.3, Sum: 26.6]
      [Update RS (ms): Min: 15.7, Avg: 17.4, Max: 17.8, Diff: 2.1, Sum: 313.1]
         [Processed Buffers: Min: 41, Avg: 54.1, Max: 77, Diff: 36, Sum: 973]
      [Scan RS (ms): Min: 0.0, Avg: 0.2, Max: 0.2, Diff: 0.2, Sum: 3.1]
      [Code Root Scanning (ms): Min: 0.0, Avg: 0.0, Max: 0.0, Diff: 0.0, Sum: 0.0]
      [Object Copy (ms): Min: 28.6, Avg: 28.7, Max: 28.8, Diff: 0.2, Sum: 516.7]
      [Termination (ms): Min: 0.0, Avg: 0.0, Max: 0.1, Diff: 0.1, Sum: 0.8]
         [Termination Attempts: Min: 4, Avg: 9.7, Max: 13, Diff: 9, Sum: 174]
      [GC Worker Other (ms): Min: 0.0, Avg: 0.1, Max: 0.2, Diff: 0.1, Sum: 1.9]
      [GC Worker Total (ms): Min: 47.7, Avg: 47.9, Max: 48.1, Diff: 0.3, Sum: 862.3]
      [GC Worker End (ms): Min: 421634488.6, Avg: 421634488.6, Max: 421634488.7, Diff: 0.1]
   [Code Root Fixup: 0.3 ms]
   [Code Root Purge: 0.0 ms]
   [Clear CT: 0.9 ms]
   [Other: 5.6 ms]
      [Choose CSet: 0.0 ms]
      [Ref Proc: 2.6 ms]
      [Ref Enq: 0.0 ms]
      [Redirty Cards: 0.2 ms]
      [Humongous Register: 0.2 ms]
      [Humongous Reclaim: 0.1 ms]
      [Free CSet: 1.4 ms]
   [Eden: 3140.0M(3140.0M)->0.0B(3524.0M) Survivors: 120.0M->160.0M Heap: 5630.7M(16.0G)->2443.0M(16.0G)]
Heap after GC invocations=5543 (full 2):
 garbage-first heap   total 16777216K, used 2501670K [0x00000003c0000000, 0x00000003c0408000, 0x00000007c0000000)
  region size 4096K, 40 young (163840K), 40 survivors (163840K)
 Metaspace       used 152940K, capacity 162166K, committed 162316K, reserved 1196032K
  class space    used 15321K, capacity 16962K, committed 17024K, reserved 1048576K
}
 [Times: user=0.87 sys=0.01, real=0.06 secs]

**Comment by karianna** [comment, marked as answer]
@kcpeppe - Any insights here?

**Reply by ZhangShushu123** [reply]
I have found the GC log can be interpreted by the PreUnifiedGCLog. The reason for the previous failure is the method I used to get the GC log.
However, I still want to ask what's the difference between PreUnifiedGCLog and UnifiedGCLog Parser.

**Reply by karianna** [reply]
The Diary detects whether the logging format is pre-unified (Java 8 or unified (Java 9+) and applies the set of parsing rules accordingly.  It's a bug if it thinks its both!

**Reply by ghost** [reply]
A Unified parser is targeted for Unified logging. The Preunified parser is for the older GC logs. The pre-unified parsers support the JDK 1.4.2 to JDK 1.8 serial, parallel, parallel new/CMS/iCMS and G1 collectors. Unified parsers support serial, parallel, parallel new/CMS , G1 and Z collectors. There is on-going work for Shenandoah that does produce events but it's very rudimentary.

As Martijn has mentioned, there is a diarization process that preparses the file to determine the version of the JDK as well as which collector is in play. The diarization process also looks for a number of other bits of metadata that can be used to describe the log.

Hope this helps

**Reply by ghost** [reply]
The log in question is JDK 8 and consequently is preunified. In this case we can see that the PrintHeapAtGC is set giving you the Heap (before|after) collection records. The data in these records is mostly redundant and the new information isn't really helpful for the purposes of tuning and as such these records are ignored.

In general, GCToolkit produced 1 event for each GC event. In all rights, the age table data belongs with the GC event but for historical reasons, this data as well as the data from safepoint events (even in cases where GC is responsible for the safe point) is treated separately. For pre-unified logs it was simpler to have a separate parsers deal with age table and safe pointing data. With Unified this is no longer the case. That said, the separation  remain in order to retain as much unity between unified and preunified as is possible.

**Reply by ZhangShushu123** [reply]
I really and sincerely appreciate for your help.
